content/client-lib-development-guide/features.textile

```diff
@@ -267,7 +267,8 @@ h3(#rest-channel). Channel
 *** @(RSL6a2)@ A set of tests must exist to ensure that the client library provides interoperability for the @extras@ field which is a JSON-encodable object (ie a value that represents a JSON @object@ value and supports serialization to and from JSON text). The test, at a minimum, should publish a message with an @extras@ object such as @{"push":[{"title":"Testing"}]}@ and ensure it is received with an equivalent JSON-encodable object
-** @(RSL6b)@ If, for example, incompatible encryption details are provided or invalid Base64 is detected in the message payload, an error message will be sent to the logger, but the message will still be delivered with last successful decoding and the @encoding@ field. For example, if a message had a decoding of "utf-8/cipher+aes-128-cbc/base64", and the payload was successfully Base64 decoded but the payload could not be encrypted because the @CipherParam@ details were not configured, the message would be delivered with a binary payload and an @encoding@ with the value "utf-8/cipher+aes-128-cbc"
+** @(RSL6b)@ If, for example, incompatible encryption details are provided or invalid Base64 is detected in the message payload, an error message will be sent to the logger, but the message will still be delivered with last successful decoding and the @encoding@ field. For example, if a message had a decoding of "utf-8/cipher+aes-128-cbc/base64", and the payload was successfully Base64 decoded but the payload could not be encrypted because the @CipherParam@ details were not configured, the message would be delivered with a binary payload and an @encoding@ with the value "utf-8/cipher+aes-128-cbc. Additional steps need to be taken if decoding failed on "vcdiff" encoding, see @(RSL6c)@ 
+** @(RSL6c)@ If decoding fails on "vcdiff" encoding that means message delta application failed.Since applying a delta depends on the previous message being present a failure in the delta application on the client side makes it impossible to apply any subsequent deltas received. That would render the channel unusable so a full message needs to be requested from Ably (the server) to replace the failed delta message. That is done by sending a new ATTACH request (the channel state might still be ATTACHED).
```

Comment:
This needs to be a lot more detailed, specific, and prescriptive. For example, think of the steps you're doing in ably-js/ably-java with `decodeFailureRecoveryInProgress` to avoid sending multiple `ATTACH`es while waiting for the previous one; the storing the channelSerial and setting it in the `ATTACH`, and so on. Those should all be spec items.

The idea is that a client library developer should, only using the spec, be able to produce a correct implementation, without reference to any existing libraries. Of course that ideal is not always achievable (and older spec items are often not nearly specific enough), but that's what we are aiming for with any new spec additions.

Reply:
Elaborated some more. Please, check it out.